### PR TITLE
Improve batch-promises codebase (typing, fixes, formatting)

### PR DESCRIPTION
**Pull request by a-ogilvie**
This PR does the following:

- Correct an error in the typescript definition of the readme
- Add a types file
- Add the correct types to `batchPromises`
- Correct an error in the test files
- Upgrade to mocha v6
- Commit a package.lock file

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,36 +13,43 @@ type Promisable<T> = T | Promise<T>;
 
 type Iterator<T, U> = (item: T) => Promisable<U>;
 
-declare function batchPromises<T, U>(batchSize: number, collection: Promisable<T[]>, callback: Iterator<T, U>) : U[];
+declare function batchPromises<T, U>(
+  batchSize: number,
+  collection: Promisable<T[]>,
+  callback: Iterator<T, U>
+): Promise<U[]>;
```
Comment (a-ogilvie) [inline README.md]: This is the correct return type for batchPromises

File: index.js [hunk 2]
```diff
@@ -1,19 +1,13 @@
...
-                });
-            }
-        })
-        .reduce(function(chain, work) {
-            return chain.then(work);
-        }, Promise.resolve([]));
-    });
+'use strict';
+
+/** @type {import('./index')} */
```
Comment (a-ogilvie) [inline index.js]: This allows type checking to work correctly.

File: test.js [hunk 3]
```diff
@@ -1,69 +1,80 @@
...
+        new Promise((resolve, reject) => {
+          setTimeout(() => {
+            if (i === 3) return reject(expectedError);
+            highestPromise = i;
+            resolve(i);
+          }, 10);
+        })
+    ).catch((actualError) => {
+      assert.equal(actualError, expectedError);
+      assert.equal(highestPromise, 2);
```
Comment (a-ogilvie) [inline test.js]: This assertion has been fixed.

**Comment by Nicktho** [comment]
Greatly appreciate this @a-ogilvie, apologies for the late response as I have been off github for a while.

**Comment by a-ogilvie** [comment]
@Nicktho No worries! Thanks for merging my code 😄

**Pull request merged**